packages/common-common/src/permissions.ts

```diff
@@ -1,63 +1,428 @@
+// Permissioning Library for Common
+// Modeled after the Discord Permissions Implementation
+
+export type Permissions = bigint;
+
+export type Permission = 'admin' | 'moderator' | 'member';
+
+export enum PermissionError {
+  NOT_PERMITTED = 'Action not permitted',
+}
+
+// Note all new actions MUST be added to the end of this list (some migrations are tied to this ordering)
 export enum Action {
-  INVITE_MEMBERS = 0,
-  BAN_MEMBERS = 1,
-  MANAGE_COMMUNITY = 2,
-  ADMINISTRATOR = 3,
-  MANAGE_ROLES = 4,
-  MANAGE_WEBHOOKS = 5,
-  MANAGE_TOPICS = 6,
-  MANAGE_CHAT_CHANNELS = 7,
-  VIEW_COMMUNITY_INSIGHTS = 8,
-  MANAGE_INVITES = 9,
-  VIEW_TOPIC = 10,
+  CREATE_CHAT = 0,
+  CREATE_REACTION = 1,
+  VIEW_REACTIONS = 2,
+  DELETE_REACTION = 3,
+  CREATE_COMMENT = 4,
+  VIEW_COMMENTS = 5,
+  EDIT_COMMENT = 6,
+  DELETE_COMMENT = 7,
+  CREATE_POLL = 8,
+  VIEW_POLLS = 9,
+  VOTE_ON_POLLS = 10,
   VIEW_CHAT_CHANNELS = 11,
   CREATE_THREAD = 12,
-  MANAGE_THREADS = 13,
-  CREATE_CHAT = 14,
-  CREATE_REACTION = 15,
-  CREATE_COMMENT = 16,
-  CREATE_POLL = 17,
-  VOTE_ON_POLLS = 18,
-  MANAGE_POLLS = 19,
+  VIEW_THREADS = 13,
+  EDIT_THREAD = 14,
+  DELETE_THREAD = 15,
+  LINK_THREAD_TO_THREAD = 16,
+  LINK_PROPOSAL_TO_THREAD = 17,
+  CREATE_TOPIC = 18,
+  VIEW_TOPICS = 19,
+  EDIT_TOPIC = 20,
+  DELETE_TOPIC = 21,
 }
 
-export type Permissions = bigint;
+const IMPLICIT_PERMISSIONS_BY_ACTION = new Map<number, Action[]>([
+  [Action.CREATE_CHAT, [Action.VIEW_CHAT_CHANNELS]],
+  [Action.VIEW_CHAT_CHANNELS, [Action.CREATE_CHAT]],
```

Comment:
I think it would be best to make this a recursively resolved tree rather than a flat hierarchy, e.g. if you can't view threads => then you can't view comments, which triggers implicit permissions on view comments, etc.

Basically, the implicit permissions should only be the DIRECT ones beneath it in the tree, rather than everything.

Reply:
I was playing with this idea as well. This makes sense I think too for long term scalability!